models/web/snowplow_media_media_stats.sql

```diff
@@ -0,0 +1,266 @@
+{{
+config(
+materialized= 'incremental',
+unique_key = 'media_id',
+sort = 'last_play',
+dist = 'media_id'
+)
+}}
+
+{% if is_incremental() %}
+
+with new_data as (
+
+  select
+    p.media_id,
+    p.duration,
+    p.media_type,
+    p.media_player_type,
+    min(case when is_played then p.start_tstamp end) as first_play,
+    max(case when is_played then p.start_tstamp end) as last_play,
+    sum(p.play_time_sec) as play_time_sec,
+    sum(case when is_played then 1 else 0 end) as plays,
+    sum(case when is_valid_play then 1 else 0 end) as valid_plays,
+    sum(case when p.is_complete_play then 1 else 0 end) as complete_plays,
+    count(distinct p.page_view_id) as impressions,
+    avg(case when is_played then coalesce(p.play_time_sec / nullif(p.duration, 0), 0) end) as avg_percent_played,
+    avg(case when is_played then p.retention_rate end) as avg_retention_rate,
+    avg(case when is_played then p.avg_playback_rate end) as avg_playback_rate
+
+from {{ ref("snowplow_media_plays_by_pageview") }} p
+
+left join {{ this }} t
+on 1 = 1
+
+where -- enough time has passed since the page_view`s start_tstamp to be able to process it a a whole (please bear in mind the late arriving data)
+p.start_tstamp < dateadd(hour, {{ var("snowplow__max_media_pv_window") }}, {{ dbt_utils.current_timestamp_in_utc() }})
+
+-- and it has not been processed yet
+and p.start_tstamp > t.last_processed_tstamp
+
+group by 1,2,3,4
+
+)
+
+, prep as (
+
+  select
+    n.media_id,
+    n.duration,
+    n.media_type,
+    n.media_player_type,
+    least(n.first_play, t.first_play) as first_play,
+    greatest(n.last_play, t.last_play) as last_play,
+    n.play_time_sec / cast(60 as {{ dbt_utils.type_float() }}) + t.play_time_min as play_time_min,
+    (n.play_time_sec / cast(60 as {{ dbt_utils.type_float() }}) + t.play_time_min)  / (n.plays + t.plays) as avg_play_time_min,
+    n.plays + t.plays as plays,
+    n.valid_plays + t.valid_plays as valid_plays,
+    n.complete_plays + t.complete_plays as complete_plays,
+    n.impressions + t.impressions  as impressions,
+    -- weighted average calculations
+    (n.avg_percent_played * n.plays / (n.plays + t.plays)) + (t.avg_percent_played * t.plays / (n.plays + t.plays)) as avg_percent_played,
+    (n.avg_retention_rate * n.plays / (n.plays + t.plays)) + (t.avg_retention_rate * t.plays / (n.plays + t.plays)) as avg_retention_rate,
+    (n.avg_playback_rate * n.plays / (n.plays + t.plays)) + (t.avg_playback_rate * t.plays / (n.plays + t.plays)) as avg_playback_rate
+
+  from new_data n
+
+  left join {{ this }} t
+  on n.media_id = t.media_id
+
+)
+
+, percent_progress_reached as (
+
+    select
+      p.media_id,
+      split_to_array(p.percent_progress_reached) as percent_progress_reached
+
+    from {{ ref("snowplow_media_plays_by_pageview") }} p
+
+    left join {{ this }} t
```

Comment:
Is this effectively a cross join?

Reply:
As clarified, this will be moved to the where clause as eventually it is only used to act as a date filter. I realised I also need to change the filter field from last_play to max(start_tstamp) from the upstream model, as this way I will also include impressions, not just plays. Otherwise any new impressions that happened in any subsequent run without any plays will be double counted.